File: Backend-node/src/controllers/textController.js
const AWS = require('aws-sdk');

// Configura AWS Rekognition
AWS.config.update({
    accessKeyId: process.env.AWS_ACCESS_KEY_ID_REKOGNITION,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY_REKOGNITION,
    region: process.env.AWS_REGION_REKOGNITION
});

const rekognition = new AWS.Rekognition();
// Controlador para extraer texto
exports.extractTextFromImage = async (req, res) => {
    try {
        // Verifica si hay un archivo de imagen
        if (!req.file) {
            return res.status(400).json({ message: 'No image file provided', status: false });
        }

        const imageBuffer = req.file.buffer;

        // Configura los parámetros para AWS Rekognition
        const params = {
            Image: {
                Bytes: imageBuffer
            }
        };

        // Llama a Rekognition para detectar texto
        rekognition.detectText(params, (err, data) => {
            if (err) {
                console.error("Error extracting text from image:", err);
                return res.status(500).json({ message: 'Error extracting text from image', error: err, status: false });
            }

            // Mapea el texto detectado
            const detectedText = data.TextDetections.map(text => text.DetectedText);
            return res.status(200).json({ message: 'Text extracted successfully', text: detectedText ,status: true });
        });
    } catch (error) {
        console.error("Error in extractTextFromImage:", error);
        res.status(500).json({ message: 'Server error', error, status: false });
    }
};

const comprehend = new AWS.Comprehend();
const translate = new AWS.Translate();

exports.detectLanguageAndTranslate = async (req, res) => {
    try {
        const { text } = req.body;

        if (!text) {
            return res.status(400).json({ message: 'No text provided' });
        }

        // Detectar el idioma del texto con Amazon Comprehend
        const detectParams = {
            TextList: [text] // Comprehend requiere una lista de textos
        };

        comprehend.batchDetectDominantLanguage(detectParams, (err, languageData) => {
            if (err) {
                console.error("Error detecting language:", err);
                return res.status(500).json({ message: 'Error detecting language', error: err });
            }

            const detectedLanguage = languageData.ResultList[0].Languages[0].LanguageCode;

            // Traducir el texto a varios idiomas (ejemplo: español, francés, alemán)
            const targetLanguages = ['en', 'fr', 'de'];
            const translations = targetLanguages.map(lang => {
                return translate.translateText({
                    SourceLanguageCode: detectedLanguage,
                    TargetLanguageCode: lang,
                    Text: text
                }).promise();
            });

            // Ejecutar las traducciones
            Promise.all(translations)
                .then(results => {
                    const translatedTexts = results.map((result, index) => ({
                        language: targetLanguages[index],
                        translatedText: result.TranslatedText
                    }));

                    return res.status(200).json({
                        message: 'Text translated successfully',
                        originalLanguage: detectedLanguage,
                        translations: translatedTexts
                    });
                })
                .catch(translationError => {
                    console.error("Error translating text:", translationError);
                    return res.status(500).json({ message: 'Error translating text', error: translationError });
                });
        });
    } catch (error) {
        console.error("Error in detectLanguageAndTranslate:", error);
        res.status(500).json({ message: 'Server error', error });
    }
};



// Función para obtener etiquetas de una imagen
exports.getLabelsFromImage = async (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ message: 'No image file provided', status: false });
        }

        // Convierte la imagen a base64
        const imageBuffer = req.file.buffer;
        const imageBase64 = imageBuffer.toString('base64');

        // Parámetros para AWS Rekognition
        const params = {
            Image: {
                Bytes: Buffer.from(imageBase64, 'base64')
            },
            MaxLabels: 10, // Máximo de etiquetas
            MinConfidence: 75 // Confianza mínima
        };

        // Llama a Rekognition para detectar etiquetas
        rekognition.detectLabels(params, (err, data) => {
            if (err) {
                console.error("Error detecting labels:", err);
                return res.status(500).json({ message: 'Error detecting labels', error: err, status: false });
            }

            // Extrae las etiquetas detectadas
            const labels = data.Labels.map(label => ({
                name: label.Name,
                confidence: label.Confidence
            }));
            console.log('Labels extracted successfully');
            return res.status(200).json({ message: 'Labels extracted successfully', labels, status: true });
        });
    } catch (error) {
        console.error("Error in getLabelsFromImage:", error);
        res.status(500).json({ message: 'Server error', error, status: false });
    }
};

File: Backend-python/src/controllers/textController.py
import os
import boto3
from flask import request, jsonify

# Configurar clientes de AWS Rekognition, Comprehend y Translate
rekognition = boto3.client('rekognition',
                           aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID_REKOGNITION'),
                           aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY_REKOGNITION'),
                           region_name=os.getenv('AWS_REGION_REKOGNITION'))

comprehend = boto3.client('comprehend',
                           aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID_REKOGNITION'),
                           aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY_REKOGNITION'),
                           region_name=os.getenv('AWS_REGION_REKOGNITION'))

translate = boto3.client('translate',
                           aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID_REKOGNITION'),
                           aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY_REKOGNITION'),
                           region_name=os.getenv('AWS_REGION_REKOGNITION'))

# Controlador para extraer texto de una imagen
def extractTextFromImage():
    try:
        # Verifica si se proporcionó una imagen
        if 'image' not in request.files:
            return jsonify({'message': 'No image file provided', 'status': False}), 400

        image_buffer = request.files['image'].read()

        # Parámetros para AWS Rekognition
        params = {
            'Image': {'Bytes': image_buffer}
        }

        # Detecta texto en la imagen
        response = rekognition.detect_text(Image=params['Image'])

        detected_text = [text['DetectedText'] for text in response['TextDetections']]
        return jsonify({'message': 'Text extracted successfully', 'text': detected_text, 'status': True}), 200

    except Exception as error:
        print("Error in extract_text_from_image:", error)
        return jsonify({'message': 'Server error', 'error': str(error), 'status': False}), 500


# Controlador para detectar el idioma y traducir el texto
def detectLanguageAndTranslate():
    try:
        text = request.json.get('text')

        if not text:
            return jsonify({'message': 'No text provided'}), 400

        # Detectar el idioma del texto
        detect_params = {'TextList': [text]}
        language_data = comprehend.batch_detect_dominant_language(TextList=detect_params['TextList'])
        detected_language = language_data['ResultList'][0]['Languages'][0]['LanguageCode']

        # Traducir el texto a varios idiomas
        target_languages = ['en', 'fr', 'de']
        translations = []

        for lang in target_languages:
            result = translate.translate_text(
                SourceLanguageCode=detected_language,
                TargetLanguageCode=lang,
                Text=text
            )
            translations.append({
                'language': lang,
                'translatedText': result['TranslatedText']
            })

        return jsonify({
            'message': 'Text translated successfully',
            'originalLanguage': detected_language,
            'translations': translations
        }), 200

    except Exception as error:
        print("Error in detect_language_and_translate:", error)
        return jsonify({'message': 'Server error', 'error': str(error)}), 500


# Controlador para obtener etiquetas de una imagen
def getLabelsFromImage():
    try:
        if 'image' not in request.files:
            return jsonify({'message': 'No image file provided', 'status': False}), 400

        image_buffer = request.files['image'].read()

        # Parámetros para AWS Rekognition
        params = {
            'Image': {'Bytes': image_buffer},
            'MaxLabels': 10,
            'MinConfidence': 75
        }

        # Detectar etiquetas
        response = rekognition.detect_labels(Image=params['Image'], MaxLabels=params['MaxLabels'], MinConfidence=params['MinConfidence'])

        labels = [{'name': label['Name'], 'confidence': label['Confidence']} for label in response['Labels']]
        return jsonify({'message': 'Labels extracted successfully', 'labels': labels, 'status': True}), 200

    except Exception as error:
        print("Error in get_labels_from_image:", error)
        return jsonify({'message': 'Server error', 'error': str(error), 'status': False}), 500
